api/internal/statemachine/statemachine.go

```diff
@@ -123,11 +162,55 @@ func (sm *stateMachine) Transition(lock Lock, nextState State) error {
 		return fmt.Errorf("invalid transition from %s to %s", sm.currentState, nextState)
 	}
 
+	fromState := sm.currentState
 	sm.currentState = nextState
 
+	// Trigger event handlers asynchronously after successful transition
+	sm.triggerHandlers(fromState, nextState)
+
 	return nil
 }
 
+func (sm *stateMachine) RegisterEventHandler(targetState State, handler EventHandler) {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+	sm.eventHandlers[targetState] = append(sm.eventHandlers[targetState], handler)
+}
+
+func (sm *stateMachine) UnregisterEventHandler(targetState State) {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+	delete(sm.eventHandlers, targetState)
+}
+
+// triggerHandlers triggers event handlers asynchronously
+func (sm *stateMachine) triggerHandlers(from, next State) {
+	handlers, exists := sm.eventHandlers[next]
+	if !exists || len(handlers) == 0 {
+		return
+	}
+	handlersCopy := make([]EventHandler, len(handlers))
+	// Make a copy of the handlers to avoid potential changes to the slice during execution
+	copy(handlersCopy, handlers)
+
+	go func(from, to State, handlerList []EventHandler) {
```

Comment:
i wonder if running in a goroutine should be an option that can be overridden / specified by the client

Reply:
We could make it so that the client can decide if they want an async or an sync handler for example. My question would be if we have an actual use for that right now or not?

Reply:
well, the reporting events we send today are actually all sync, not async, no?

Reply:
i think i agree with salah here that it should be sync. async is hard because we may need to catch signals to ensure that all handlers execute before the process exits.

Reply:
i think i agree with salah here that it should be synchronous as it is one more thing to gracefully shutdown before the process exits

Reply:
happy to make them sync in that case, we just need to be mindful that wtv we hook up to these handlers in the future will be sync and will mean blocking the transition while it happens.

Reply:
Done via https://github.com/replicatedhq/embedded-cluster/pull/2363/commits/9581235c330930803d2ce6a96f287a810013739a